common/events/events_io.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/events_io.hpp>
+
+namespace mender {
+namespace common {
+namespace events {
+namespace io {
+
+AsyncReaderFromReader::AsyncReaderFromReader(
+	EventLoop &loop, mender::common::io::ReaderPtr reader) :
+	cancelled_(make_shared<atomic<bool>>(false)),
+	reader_(reader),
+	loop_(loop) {
+}
+
+AsyncReaderFromReader::~AsyncReaderFromReader() {
+	Cancel();
+}
+
+error::Error AsyncReaderFromReader::AsyncRead(
+	vector<uint8_t>::iterator start,
+	vector<uint8_t>::iterator end,
+	mender::common::io::AsyncIoHandler handler) {
+	if (reader_thread_.joinable()) {
+		// Clean up previous operation. Careful, this can hang if AsyncRead is called twice
```

Comment:
Can we not detect if the handler has been run?

Reply:
I did even better actually: I moved the joining to the handler, so if it's joinable when we get here, it's an error which we can immediately report.